device/sycl/src/utils/barrier.hpp

```diff
@@ -8,40 +8,59 @@
 #pragma once
 
 // Project include(s).
-#include "traccc/definitions/qualifiers.hpp"
+#include "traccc/device/concepts/barrier.hpp"
 
-// SYCL includes
+// SYCL include(s).
 #include <sycl/sycl.hpp>
 
-namespace traccc::sycl {
+namespace traccc::sycl::details {
 
+/// A barrier type for SYCL kernels.
+template <std::size_t DIMENSIONS = 1>
 struct barrier {
-    barrier(::sycl::nd_item<1> item) : m_item(item) {}
 
-    TRACCC_DEVICE
-    void blockBarrier() { m_item.barrier(); }
+    /// Verify that the barrier is used on a supported dimensionality.
+    static_assert(DIMENSIONS > 0 && DIMENSIONS < 4,
+                  "Only 1D, 2D, and 3D barriers are supported.");
 
-    TRACCC_DEVICE
-    bool blockAnd(bool predicate) {
-        m_item.barrier();
+    /// Dimensions of the barrier
+    static constexpr std::size_t dimensions = DIMENSIONS;
+
+    /// Constructor
+    ///
+    /// @param item The @c ::sycl::nd_item<1> to use for the barrier.
+    ///
+    explicit barrier(const ::sycl::nd_item<dimensions>& item) : m_item(item) {}
+
+    inline void blockBarrier() const {
+        ::sycl::group_barrier(m_item.get_group());
+    }
```

Comment:
Wondering if we should accept `m_item.get_group()` as the constructor argument, then we don't need to template the barrier on the number of dimensions. :thinking:

Reply:
I thought about that for a while. The reason I didn't go that way is that all these functions receive templated group objects as arguments, by value. We could mimic that in this class, making it into a generic template having a `GROUP` member. :thinking:

Holding on to the "more concrete" item object just seemed a bit more robust against SYCL implementation changes. :thinking:

Reply:
Yeah alright, makes sense; it is rather unfortunate to have to template this class but sure. :+1: